test/suite/test_checkpoint08.py

```diff
@@ -0,0 +1,119 @@
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+# test_checkpoint08.py
+# Test that the checkpoints doesn't discard data that is older than the checkpoint's oldest timestamp.
+
+import threading, time
+import wiredtiger, wttest
+from wtthread import checkpoint_thread
+
+def timestamp_str(t):
+    return '%x' % t
+
+class test_checkpoint08(wttest.WiredTigerTestCase):
+    conn_config = 'cache_size=5MB,timing_stress_for_test=[checkpoint_slow]'
+    session_config = 'isolation=snapshot'
+
+    def large_updates(self, uri, value, nrows, commit_ts):
+        # Update a large number of records.
+        session = self.session
+        cursor = session.open_cursor(uri)
+        for i in range(0, nrows):
+            session.begin_transaction()
+            cursor[str(i)] = value
+            session.commit_transaction('commit_timestamp=' + timestamp_str(commit_ts))
+        cursor.close()
+
+    def large_deletes(self, uri, nrows, commit_ts):
+        # Delete a large number of records.
+        session = self.session
+        cursor = session.open_cursor(uri)
+        for i in range(0, nrows):
+            session.begin_transaction()
+            cursor.set_key(str(i))
+            cursor.remove()
+            session.commit_transaction('commit_timestamp=' + timestamp_str(commit_ts))
+        cursor.close()
+
+    def check(self, check_value, uri, nrows, read_ts):
+        session = self.session
+        session.begin_transaction('read_timestamp=' + timestamp_str(read_ts))
+        cursor = session.open_cursor(uri)
+        count = 0
+        for k, v in cursor:
+            self.assertEqual(v, check_value)
+            count += 1
+        session.commit_transaction()
+        self.assertEqual(count, nrows)
+
+    def test_checkpoint08(self):
+        nrows = 1000
+
+        uri = "table:checkpoint08"
+        self.session.create(uri, 'key_format=S,value_format=S')
+
+        value1 = "a" * 500
+        value2 = "b" * 500
+
+        # Pin oldest and stable to timestamp 10.
+        self.conn.set_timestamp('oldest_timestamp=' + timestamp_str(10) +
+            ',stable_timestamp=' + timestamp_str(10))
+
+        # Insert keys at timestamp 20
+        self.large_updates(uri, value1, nrows, 20)
+
+        # Update the keys at timestamp 30
+        self.large_updates(uri, value2, nrows, 30)
+
+        # Remove the keys at timestamp 40
+        self.large_deletes(uri, nrows, 40)
+
+        # Set stable to timestamp 50.
+        self.conn.set_timestamp('stable_timestamp=' + timestamp_str(50))
+
+        # Create a checkpoint thread
+        done = threading.Event()
+        ckpt = checkpoint_thread(self.conn, done)
+        try:
+            ckpt.start()
+            # Wait until checkpoint has get its timestamps
+            time.sleep(1)
```

Comment:
I'm worried that this is not very reliable but I don't have a better solution.

Reply:
Your thought and comment is a good indicator that maybe a `test/csuite` program is the more natural place for a test instead of python. I suggest looking at `test/csuite/wt3363_checkpoint_op_races/main.c` for a test that might be a good starting place that should already have the infrastructure you need for this issue.

Reply:
Thanks Sue. I'll explore that.